src/transformers/models/groupvit/modeling_groupvit.py

```diff
@@ -0,0 +1,1864 @@
+    ), f"{attentions.shape[2]} = {feat_height} x {feat_width}, height={height}, width={width}"
+
+    batch_size = attentions.shape[0]
+    groups = attentions.shape[1]  # number of group token
+    # [batch_size, groups, height*width, groups] -> [batch_size, groups, height, width]
+    attentions = attentions.reshape(batch_size, groups, feat_height, feat_width)
+    attentions = nn.functional.interpolate(
+        attentions, size=(height, width), mode="bilinear", align_corners=align_corners
+    )
+    return attentions
+
+
+def get_grouping_from_attentions(attentions, hw_shape):
+    """
+    Args:
+        attentions (`tuple(torch.FloatTensor)`: tuple of attention maps returned by `GroupViTVisionTransformer`
+        hw_shape (`tuple(int)`): height and width of the output attention map
+    Returns:
+        `torch.Tensor`: the attention map of shape [batch_size, groups, height, width]
+    """
+
+    attn_maps = []
+    with torch.no_grad():
+        prev_attn_masks = None
+        for attn_masks in attentions:
+            # [batch_size, num_groups, height x width] -> [batch_size, height x width, num_groups]
+            attn_masks = attn_masks.permute(0, 2, 1).contiguous()
+            if prev_attn_masks is None:
+                prev_attn_masks = attn_masks
+            else:
+                prev_attn_masks = prev_attn_masks @ attn_masks
+            # [B, HxW, G] -> [B, G, HxW] -> [B, G, H, W]
+            cur_attn_map = resize_attn_map(prev_attn_masks.permute(0, 2, 1).contiguous(), *hw_shape)
+            attn_maps.append(cur_attn_map)
+
+    # [B, G, H, W]
+    final_grouping = attn_maps[-1]
+
+    return final_grouping
+
+
+class GroupViTAttention(nn.Module):
+    def __init__(self, config: GroupViTVisionConfig):
+        super().__init__()
+        self.num_heads = config.num_attention_heads
+        head_dim = config.hidden_size // self.num_heads
+        self.scale = head_dim**-0.5
+
+        self.q_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.k_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.v_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.proj = nn.Linear(config.hidden_size, config.hidden_size)
+
+    def forward(self, query, key=None):
+        batch_size, query_length, channels = query.shape
+        if key is None:
+            key = query
+        key_length = key.size(1)
+
+        # [batch_size, num_heads, query_length, channels//num_heads]
+        q = (
+            self.q_proj(query)
+            .reshape(batch_size, query_length, self.num_heads, channels // self.num_heads)
+            .permute(0, 2, 1, 3)
+        )
+        # [batch_size, num_heads, key_length, channels//num_heads]
+        k = (
+            self.k_proj(key)
+            .reshape(batch_size, key_length, self.num_heads, channels // self.num_heads)
+            .permute(0, 2, 1, 3)
+        )
+        # [batch_size, num_heads, key_length, channels//num_heads]
+        v = (
```

Comment:
We usually try to avoid one-letter variable names. `query`, `key` and `value` instead?